### ISSUE TITLE
Asymmetric encryption for "secret" config values.

### ISSUE THREAD

**Issue by ellismg** [pulumi/pulumi#2587]
Today, when you encrypt a configuration value, the plain-text is encrypted using a service managed key (when using the local backend, the key is derived from the passphrase).  While we plan to long term support pluggable backends for managing this encryption (see https://github.com/pulumi/pulumi/issues/1547) we've had a request for supporting a [built in asymmetric encryption like thing](https://github.com/pulumi/pulumi/issues/1547#issuecomment-468834825).

While you could imagine not doing this work and just doing the larger item described in 1547, we think having this as a default built in will unlock a bunch of interesting use cases and gives us like 80% of the value of the above.  We'll use the design and implementation of this feature to start thinking about the bigger picture while still delivering something usable.

**Comment by AaronFriel** [comment]
FWIW, this is lower priority for us than #1547 by far. We can hack our way to asymmetric encryption of secrets for developers by using KMS.

Pluggable backends for secrets and not transmitting secrets to the Pulumi SaaS backend are much, much higher priority.

**Comment by joeduffy** [comment]
The intent here is that the asymmetric encryption option would be the first pluggable backend offered. That is to say, we plan to use this work item to put in place the capability to select a backend, of which there would only be two options:

* `auto`: the current option that uses an automatically managed KMS key plus per-stack AESGCM256 key managed by the Pulumi SaaS (same as today)
* `aesgcm256-local`: an option to use a client-managed encryption key so that plaintext is never shared with the Pulumi SaaS, and so it never has complete enough information to decrypt ciphertext

@AaronFriel I'm curious if the `aesgcm256-local` mode would unblock you for the time being. Our plan is most certainly to keep chugging immediately afterwards with #1547. I expect the third backend to be `kms`, which leverages your own account managed KMS key.

**Comment by joeduffy** [comment]
These are out for review, but I am pushing to M23 so we can close the prior sprint.

**Comment by ellismg** [comment]
The ability to use passphrase based encryption for service managed landed in 0.17.11
